src/Octoshift/CodeScanningAlertService.cs

```diff
@@ -0,0 +1,209 @@
+            _sourceGithubApi = sourceGithubApi;
+            _targetGithubApi = targetGithubApi;
+            _log = octoLogger;
+        }
+
+        public virtual async Task MigrateCodeScanningAlerts(string sourceOrg, string sourceRepo, string targetOrg,
+            string targetRepo, bool dryRun)
+        {
+            var defaultBranch = await _sourceGithubApi.GetDefaultBranch(sourceOrg, sourceRepo);
+            _log.LogInformation($"Found default branch: {defaultBranch} - migrating code scanning alerts only of this branch.");
+            await MigrateAnalyses(sourceOrg, sourceRepo, targetOrg, targetRepo, defaultBranch, dryRun);
+            await MigrateAlerts(sourceOrg, sourceRepo, targetOrg, targetRepo, defaultBranch, dryRun);
+        }
+
+        protected internal virtual async Task MigrateAnalyses(string sourceOrg, string sourceRepo, string targetOrg,
+            string targetRepo, string branch, bool dryRun)
+        {
+            _log.LogInformation($"Migrating Code Scanning Analyses from '{sourceOrg}/{sourceRepo}' to '{targetOrg}/{targetRepo}'");
+
+            var analyses = await _sourceGithubApi.GetCodeScanningAnalysisForRepository(sourceOrg, sourceRepo, branch);
+            analyses = analyses.ToList();
+
+            var successCount = 0;
+            var errorCount = 0;
+
+            _log.LogVerbose($"Found {analyses.Count()} analyses to migrate.");
+
+            if (dryRun)
+            {
+                _log.LogInformation($"Running in dry-run mode. The following Sarif-Reports would now be downloaded from '{sourceOrg}/{sourceRepo}' and then uploaded to '{targetOrg}/{targetRepo}':");
+                foreach (var analysis in analyses)
+                {
+                    _log.LogInformation($"    Report of Analysis with Id '{analysis.Id}' created at {analysis.CreatedAt}.");
+                }
+                return;
+            }
+
+            foreach (var analysis in analyses)
+            {
+                var sarifReport = await _sourceGithubApi.GetSarifReport(sourceOrg, sourceRepo, analysis.Id);
+                _log.LogVerbose($"Downloaded SARIF report for analysis {analysis.Id}");
+                try
+                {
+                    await _targetGithubApi.UploadSarifReport(targetOrg, targetRepo,
+                        new SarifContainer
+                        {
+                            Sarif = sarifReport,
+                            Ref = analysis.Ref,
+                            CommitSha = analysis.CommitSha
+                        });
+                    _log.LogInformation($"Successfully Migrated report for analysis {analysis.Id}");
+                    ++successCount;
+                }
+                catch (HttpRequestException httpException)
+                {
+                    if (httpException.StatusCode.Equals(HttpStatusCode.NotFound))
+                    {
+                        _log.LogVerbose($"No commit found on target. Skipping Analysis {analysis.Id}");
```

Comment:
Is a missing commit the only reason why this could 404? If we're not certain, then I think it would be better to be a bit less certain in this log.

Reply:
Well spotted - a 404 can also occur if the target token is lacking permissions.

I added this to the log and also made it a `warn` rather than only when being in `verbose` mode.

This was originally in there when I was still allowing pull-request analysis to be migrated - there it could happen that commits were missing as, during an action run, merge-commits were created on the runner itself that never got pushed back to GitHub. In that case, a skipped analysis was no problem at all.

But not having it only for the main branch, it actually is.